File: python/3081. Replace Question Marks in String to Minimize Its Value.py
class Solution:
    def minimizeStringValue(self, s: str) -> str:
        freq = [0] * 26
        count = 0
        for c in s:
            if c == '?':
                count += 1
            else:
                freq[ord(c) - ord('a')] += 1
        arr = []
        def find(freq):
            min = 1000000000
            minFreqChar = '?'
            for i in range(26):
                if freq[i] < min:
                    min = freq[i]
                    minFreqChar = chr(i + ord('a'))
            freq[ord(minFreqChar) - ord('a')] += 1
            return minFreqChar
        for i in range(count):
            arr.append(find(freq))
        arr = sorted(arr)
        index = 0
        answer = []
        for c in s:
            if c == '?':
                answer.append(arr[index])
                index += 1
            else:
                answer.append(c)
        return ''.join(answer)

File: java/3081. Replace Question Marks in String to Minimize Its Value.java
class Solution {
    public String minimizeStringValue(String s) {
        int[] freq = new int[26];
        int count = 0;
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) != '?') {
                freq[s.charAt(i)-'a']++;
            } else {
                count++;
            }
        }
        char[] arr = new char[count];
        for (int i = 0; i < count; i++) {
            arr[i] = find(freq);
        }
        Arrays.sort(arr);
        StringBuilder sb = new StringBuilder();
        int index = 0;
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '?') {
                sb.append(arr[index++]);
            } else {
                sb.append(s.charAt(i));
            }
        }
        return sb.toString();
    }
    private char find(int[] freq) {
        int minFreq = Integer.MAX_VALUE;
        char minFreqC = '?';
        for (int i = 0; i < freq.length; i++) {
            if (freq[i] < minFreq) {
                minFreq = freq[i];
                minFreqC = (char)(i + (int)'a');
            }
        }
        freq[minFreqC-'a']++;
        return minFreqC;
    }
}